### DISCUSSION TITLE
Running telegram-chat-forward on Python Anywhere

### DISCUSSION THREAD

**Discussion by aahnik** [aahnik/tgcf#23] [Q&A] [answered]
How can this script be run on Python Anywhere?

https://www.pythonanywhere.com/

- Free
- No hassle of downloading and configuring `python` or `git` or `pip`, everything is pre-installed
- You get a full-fledged bash terminal
- You could close the screen, and the script will still be running until it finishes the work (usefully if you have lots of forwards and it takes a lot of time)
- Has a web-based file editor, where you can edit the `.env` or `config.ini`
- You can schedule a task to run periodically

**Comment by aahnik** [comment, marked as answer]
1. First of all go to https://www.pythonanywhere.com/ and create an account there
2. Open you dashboard, you see this on left:
   
![image](https://user-images.githubusercontent.com/66209958/109107767-59104d00-7758-11eb-8d7f-f904a55acfff.png)

3. Click on " $ Bash " below "New Console"
4. A Linux terminal will open like this
 
![image](https://user-images.githubusercontent.com/66209958/109107885-8bba4580-7758-11eb-96ca-e66c99937842.png)

5. Now follow the steps, from https://github.com/aahnik/telegram-chat-forward/#installation

------


How to create files?

1. Click on Files
![image](https://user-images.githubusercontent.com/66209958/109110196-a989a980-775c-11eb-9b23-e5abc8946c18.png)

2. In the left, go the **telegram-chat-forward** directory (folder)
![image](https://user-images.githubusercontent.com/66209958/109110243-bf976a00-775c-11eb-9f2b-d0beeae67f56.png)

3. You will see this
![image](https://user-images.githubusercontent.com/66209958/109110284-d3db6700-775c-11eb-9c06-68fd04d8e431.png)

**Comment by thebeginning1** [comment]
Hi, I tried to run the script on pythonanywhere like a week ago but it didn't work so I did some research and I found out that pythonanywhere doesn't support Mtproto for free accounts.

**Comment by sbahlouli** [comment]
hello, I'm having trouble installing the requirements.

![image](https://user-images.githubusercontent.com/66209958/109909392-0a206580-7ccc-11eb-8e4e-e832b17a1c64.png)

**Reply by aahnik** [reply]
make sure you use the latest version of python and pip, and also a virtual environment. if the error still persists, don't worry, and go ahead.

**Comment by anymeofu** [comment]
Can You Make a repl.it version ? Because everyone use that.....

**Reply by aahnik** [reply]
there is nothing called a repl.it version. You can use the code anywhere.

Go to repl.it , clone the repo, and use it

**Reply by tissole** [reply]
On repl.it free tier the project becomes public by default. Replit hides the env file but forwarder.session is visible to everyone.

**Comment by tissole** [comment]
Actually, I think the script can run on pythonanywhere.com on the free tier but with some adjustments. On their [page](https://help.pythonanywhere.com/pages/403ForbiddenError/) they say that is possible through their proxy.  Can this workaround work with the script? How to use it?

**Reply by aahnik** [reply]
wow, thanks for this brilliant suggestion. I will learn more about this and implement the required changes to support running on python anywhere free tier.

**Reply by tissole** [reply]
I found [this](https://github.com/lstil/pythonanywhere-telegram-bot), maybe can help you.

**Reply by aahnik** [reply]
thanks for your efforts, but here we don't use webhook or polling. I am using Telethon, which the python mtproto library.

**Comment by yhsomani** [comment]
Can You help Me to fix this problem 
![problem](https://user-images.githubusercontent.com/57716157/113776938-158e0300-9748-11eb-8c99-cfe1f23c2ff6.png)

**Reply by aahnik** [reply]
read this https://github.com/aahnik/tgcf/discussions/23#discussioncomment-410020

**Reply by jerung3000** [reply]
hi aahnik..so this method cannot be used on python anywhere?

Because when I run the script "python3 forwarder.py", i get the same error as yhsomani..

..

**Reply by aahnik** [reply]
if you have a paid account of python anywhere, it will support all connections. in a free account, it blacklists certain connections. but can be overcome by using proxy (will update you about that later)

**Reply by ankit-sinha-18** [reply]
I am using this repo on Heroku on free dynos
![Screenshot_2021_0414_180056](https://user-images.githubusercontent.com/72637068/114812701-24bb2380-9dce-11eb-98cd-4f51b8c6a2d6.gif)
😊

**Reply by ankit-sinha-18** [reply]
@aahnik Are you not active on Telegram?
I have sent you message in pm.

**Reply by aahnik** [reply]
you will soon get buttons to one click deploy to heroku, digital ocean, gitpod, reply, google cloud etc.

i am working towards the best user experience.

this is basically python code, and you can run anywhere.

but problem with heroku is that, the config.ini is not saved. the next time you forward it will start from beginning.

heroku has a ephimeral file system.

**Reply by ankit-sinha-18** [reply]
Yes, heroku doesn't save config.ini
But it is suitable when files are in large number.
Just deploy and it will run for 24 hours

After that ,you have to set offset manually and run it again

**Reply by aahnik** [reply]
true. this kind of thing can  be also run using a github action. an workflow is allowed to run for 72 hours.

github actions also support cron syntax. ie, you can run periodically. example, once every day, or one time per week, or on every Friday, whatever you wish.

github actions, can also save the (config) file, in the github repo. the `.env` stuff like API id, and API has, can be saved in github secrets.

github actions are free for public repos. and also has lot of free action minutes for private onces.

i will soon publish a github action , you can use to run the job periodically.

**Reply by ankit-sinha-18** [reply]
Woow, that's great 👍

**Reply by aahnik** [reply]
@ankit-sinha-18 please star this repo, and share with your friends, and other potential users. help it reach a hundred stars.

i already have the code for lot of features, will release everything together, (as old way of using will break, and new easy way will be there).

thank you.

**Reply by ankit-sinha-18** [reply]
I have already starred this repo,
And I have shared this repo link in my Telegram Channel also

**Reply by jerung3000** [reply]
Hi Aahnik..

I tried this repo in python on windows and get this syntaxerror..

```shell
_C:\telegram-chat-forward>python forwarder.py
  File "C:\telegram-chat-forward\forwarder.py", line 19
    def intify(string)
                      ^
SyntaxError: invalid syntax_
```

I already tried to installed dependancies and only can installed this:
```shell
pip3 install -r requirements.txt
```

But this file cannot be installed :
```shell
python3 -m venv venv && source venv/bin/activate
```

My question was can this repo be run on stand PC with python?

How to installed dependancies on windows?

**Reply by aahnik** [reply]
you are most probably using older version of python, or you messed up something.

Dont, worry, today I am releasing a new version. Will write more about that soon. 

Join the the channel [@tg_cf](https://telegram.me/tg_cf) to stay updated.
